src/store/fs/util/entity_manager.rs

```diff
@@ -0,0 +1,1322 @@
+
+use n0_future::{future, FuturesUnordered};
+use tokio::sync::{mpsc, oneshot};
+
+/// Trait to reset an entity state in place.
+///
+/// In many cases this is just assigning the default value, but e.g. for an
+/// `Arc<Mutex<T>>` resetting to the default value means an allocation, whereas
+/// reset can be done without.
+pub trait Reset: Default {
+    /// Reset the state to its default value.
+    fn reset(&mut self);
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ShutdownCause {
+    /// The entity is shutting down gracefully because the entity is idle.
+    Idle,
+    /// The entity is shutting down because the entity manager is shutting down.
+    Soft,
+    /// The entity is shutting down because the sender was dropped.
+    Drop,
+}
+
+/// Parameters for the entity manager system.
+pub trait Params: Send + Sync + 'static {
+    /// Entity id type.
+    ///
+    /// This does not require Copy to allow for more complex types, such as `String`,
+    /// but you have to make sure that ids are small and cheap to clone, since they are
+    /// used as keys in maps.
+    type EntityId: Debug + Hash + Eq + Clone + Send + Sync + 'static;
+    /// Global state type.
+    ///
+    /// This is passed into all entity actors. It also needs to be cheap handle.
+    /// If you don't need it, just set it to `()`.
+    type GlobalState: Debug + Clone + Send + Sync + 'static;
+    /// Entity state type.
+    ///
+    /// This is the actual distinct per-entity state. This needs to implement
+    /// `Default` and a matching `Reset`. It also needs to implement `Clone`
+    /// since we unfortunately need to pass an owned copy of the state to the
+    /// callback - otherwise we run into some rust lifetime limitations
+    /// <https://github.com/rust-lang/rust/issues/100013>.
+    ///
+    /// Frequently this is an `Arc<Mutex<T>>` or similar. Note that per entity
+    /// access is concurrent but not parallel, so you can use a more efficient
+    /// synchronization primitive like [`AtomicRefCell`](https://crates.io/crates/atomic_refcell) if you want to.
+    type EntityState: Default + Debug + Reset + Clone + Send + Sync + 'static;
+    /// Function being called when an entity actor is shutting down.
+    fn on_shutdown(
+        state: entity_actor::State<Self>,
+        cause: ShutdownCause,
+    ) -> impl Future<Output = ()> + Send + 'static
+    where
+        Self: Sized;
+}
+
+/// Sent to the main actor and then delegated to the entity actor to spawn a new task.
+pub(crate) struct Spawn<P: Params> {
+    id: P::EntityId,
+    f: Box<dyn FnOnce(SpawnArg<P>) -> future::Boxed<()> + Send>,
+}
+
+pub(crate) struct EntityShutdown;
+
+/// Argument for the `EntityManager::spawn` function.
+pub enum SpawnArg<P: Params> {
+    /// The entity is active, and we were able to spawn a task.
+    Active(ActiveEntityState<P>),
+    /// The entity is busy and cannot spawn a new task.
+    Busy,
+    /// The entity is dead.
+    Dead,
+}
```

Comment:
Two somewhat related comments:
1. Given we always ignore `Busy` and `Dead` in our implementation right now and both result in `HashSpecificCommand::on_error`, perhaps we can remove `SpawnArg` altogether and just return an error from `.spawn`/`.spawn_blocking` in the main actor? Then these functions wouldn't have to be `async` anymore, too.
2. `Busy` is misleading I think. An entity actor can be "busy" in the sense that it does work, but is not yet at 100% capacity. Perhaps another name for this could be `Overloaded`? Or `AtCapacity`?

Reply:
I like having the FnOnce being called exactly once and not at most once. Sadly we are already in callback land, so might as well embrace it. It is a matter of taste I guess.

Reply:
Not sure what to do about Busy. It is almost impossible to reach this. This happens if you send messages faster than they can be converted into local tasks on the recv side. I guess a burst of tasks would do it. Maybe just call it QueueFull? But then this does not really appear in the public API, so I don't lose sleep about it.